338 manage ws members

reinhardt opened the pull request:
UI for managing workspace members and their roles.
- add multiple members
- remove members
- assign/change roles
- remove roles
- display roles
- hide functions for unprivileged users

Compare with prototype.

reinhardt committed:
Connected button to remove members, allowed adding multiple members.
refs #338 #11988

reinhardt committed:
Implemented UI for managing workspace members and their roles.
This incorporates recent prototype changes. Refs #338 #11988

gyst commented:
Lots of security-sensitive feature work, but no test coverage at all. Please add some robot test coverage that:
- validates the extra roles visually
- but also validates the extra permissions actually, for example setting a workspace to 'consume', then elevating a single user, then validate that this user has gained 'add' permission

See for example the work I did in
https://github.com/ploneintranet/ploneintranet/commit/347a99d7b4bc750ce38f9de866377d12bf98e81f

reinhardt commented:
I'm not sure the second part is within the scope of this. I just connected UI elements to existing role management features. The latter may or may not be tested - I didn't touch them.

That said, robot tests that verify the UI are still to do, as mentioned in #338.

gyst commented:
@reinhardt You don't have to do heavy checks. Just something that not tautologically tests your own changes only. If you're changing permissions, other elements in the UI should change. Anchoring the test on one of those changes is sufficient.

gyst commented:
@reinhardt please also merge in master so I can auto-merge on github once this is complete.

reinhardt committed:
handling of groups as workspace members, refs #338 #11988

reinhardt committed:
fixed flake8 failure

reinhardt committed:
added robot test
'The manager can grant the Producer role in a Cosumer workspace'
refs #338 #11988

reinhardt committed:
added robot test
'The manager can remove a workspace member'
refs #338 #11988

reinhardt committed:
Merge branch 'master' into 338-manage-ws-members

Conflicts:
	src/ploneintranet/workspace/browser/tiles/sidebar.py

reinhardt commented:
Added two robot tests and rebased/merged master.

reinhardt committed:
added two robot tests
'The manager can remove a special role from a workspace member'
'The manager can change a special role'
refs #338 #11988

reinhardt commented:
...and two more.

gyst commented on src/ploneintranet/workspace/browser/tiles/sidebar.py:
```diff
@@ -127,22 +136,60 @@ def users(self):
+                    default_role = ws.participant_policy.title()
+                    groups = [default_role]
+                    if role:
+                        groups.append(role)
```
This is in line with our original technical design, but it conflicts with the UI design. This adds custom roles on top of the default workspace roles. The UI clearly indicates that you're reassigning the user from the default role to another role.

In this implementation, if you're in a 'publish' workspace you can move users to 'consume' visually but it has no security effect whatsoever, because the user is in both ('publish', 'consume') groups and publish is a superset of consume.

The reason that this problem doesn't show up in the tests, is because they are run in a consume workspace and elevate permissions.

To summarize: with this implementation you can only elevate permissions, not assign lesser permissions than the default. IMO that should be changed so that a custom role _replaces_ rather than _adds to_ the default group.

@mattss do you agree?

reinhardt replied:
> The UI clearly indicates that you're reassigning the user from the default role to another role.

Interesting - to me it doesn't. The fact that some users have this orange bubble with a role while regular users don't have a bubble to me conveys the idea that these users have something extra on top of what the other users have. If all other users had a bubble with the default role, maybe in a different colour, that would in my eyes be closer to what you describe.

gyst replied:
I get your point, visually. However the wording of the interactions makes it clear this is about assigning roles, not assigning _extra_ roles. Probably fixable by replacing the default group with the new one, instead of setting both.

reinhardt replied:
> However the wording of the interactions makes it clear this is about assigning roles, not assigning extra roles. 

True.

> Probably fixable by replacing the default group with the new one, instead of setting both.

Will do (assuming @matss doesn't veto).

mattss replied:
I _think_ I just broke the ability for policy exceptions in a recent commit (https://github.com/ploneintranet/ploneintranet/commit/06ac676d57546a66e99d389475f3ba2da61dc6c8) (sorry!)

Previously, this only moved uses from the currently policy to the new policy, but I changed it to move _all_ members to the new policy, because we were having persistence issues. I now think this move was foolish, so we may need to switch back to the logic where it only moves users who were in the old policy group (see diff for what it was doing originally).

If you want me to try and fix that, please let me know.

gyst replied:
I think the new code is much more robust, and, as I note, it's also in line with how Cornelis has structured the UI. Simplicity is our mantra. Please keep the current behavior and let's do "replace" instead of "add" group roles.

gyst replied:
Oops hold. I now see our testcontent does set multiple roles. Lemme think.

gyst commented on src/ploneintranet/suite/tests/lib/keywords.robot:
```diff
@@ -392,6 +439,11 @@ I cannot see the document
 
+Allan has the option to create a document
+    I am logged in as the user allan_neece
+    I go to the Shareholder Information Workspace
```
This is a 'consume' workspace which is the lowest privilege level. That's OK for testing privilege enhancement, but not for reducing access rights. For that you'd need a different workspace with a higher default policy. See other comment.

gyst commented:
@reinhardt This is great work, thanks for the care you put in finishing and testing. I'm sorry it is your bad luck to expose a design inconsistency that I should have caught earlier.

reinhardt commented:
Don't worry about it, always happy to stir things up! ;-) And thanks for the appreciation.

gyst commented:
So:
- make roles replacing not additive
- fix the collective.workspace permission problem
- also show 'Team Manager' as a role

gyst commented:
@reinhardt Matt will take over now.

reinhardt commented:
Roger that.

mattss committed:
Ensure we only update relevant users on participation policy change

Only move users to a new default participation policy group if they were in the old group. This allows support for 'exception' users who were previously moved out of the default group.

mattss committed:
Tests for policy exceptions

mattss committed:
Replace existing groups on role change

mattss committed:
Better docstring for this key method

mattss committed:
Ignore 'Members' automatic group when adding to team

mattss committed:
Allow admin role to be assigned.

mattss committed:
Robot test for assigning admins

mattss committed:
Merge branch 'master' into 338-manage-ws-members

gyst committed:
test also rights removal

gyst commented:
Great work all!

gyst committed:
Merge branch 'master' into 338-manage-ws-members

Conflicts:
	src/ploneintranet/workspace/browser/tiles/sidebar.py

gyst commented:
RRgh the merge broke the test. Will fix that. Later.

pilz committed:
Merge remote-tracking branch 'origin/master' into 338-manage-ws-members

pilz commented:
I am checking the tests now

gyst committed:
Sjeez. Spent hours on robot timing issues.

gyst committed:
try to avoid StaleElementReferenceException heisenbug

gyst commented:
@pilz I fixed this!

pilz commented:
lol :) Yes, I found no errors in the tests ;) Super!
